### ISSUE TITLE
csgraph.maximum_flow can cause Python crash for large but very sparse matrices

### ISSUE THREAD

**Issue by fuglede** [scipy/scipy#14385]
While reviewing #14358, I ran into the following issue (which affects both the implementation in that PR and the currently available SciPy implementation):

#### Reproducing code example:

Consider the flows calculation of the following very simple graph (in which each vertex has exactly one successor), represented by the matrix with ones on the superdiagonal and zeros elsewhere:

```python
import numpy as np
from scipy.sparse import csr_matrix
from scipy.sparse.csgraph import maximum_flow

def calc_flow(n):
    indices = np.arange(1, n)
    indptr = np.array(list(range(n)) + [n - 1])
    data = np.ones(n - 1, dtype=int)
    A = csr_matrix((data, indices, indptr), shape=(n, n))
    return maximum_flow(A, 0, n - 1)
```

On my machine, running this for `n in (10, 100, 1_000, 10_000)`, I get the expect result (a flow of 1), but for `n = 100_000`, Python crashes. This seems to suggest faulty memory allocation somewhere in the preprocessing for the flow calculation; the graph itself is so small that performing the calculation should be unproblematic.

#### Scipy/Numpy/Python version information:

```
1.6.1 1.19.2 sys.version_info(major=3, minor=7, micro=6, releaselevel='final', serial=0)
```

**Comment by fuglede** [comment]
The issue is due to an overflow in at `haystack = rows*n + cols` in `_flow.pyx`. Replacing both instances of

```python
rows = acoo.row
cols = acoo.col
```
with
```python
rows = acoo.row.astype(np.uint64)
cols = acoo.col.astype(np.uint64)
```

and
```python
return rev_edge_ptr, rows
```
with
```python
return rev_edge_ptr, rows.astype(ITYPE)
```

is enough to take care of the problem, or at least, make it possible to run much larger matrices through the solver.

This is not a particularly nice fix though. And while both of `_add_reverse_edges` and `_make_edge_pointers` are nice and NumPy'y, they're not particularly performant (with concrete improvements having been suggested on the mailing list), and with the work in #14358, they end up being actual bottlenecks for the full implementation, so a better solution would be to simply rethink and rewrite those two functions.

**Comment by tupui** [comment]
Thanks @fuglede. The fix being simple, would it still make sense to do it until you/someone rewrite the functions?

**Comment by fuglede** [comment]
The mailing list thread I had in mind was [this one](https://mail.python.org/pipermail/scipy-dev/2020-January/023914.html); it'd need some tweaking (I'd assume, for instance, a priori that it wouldn't quite be enough if some backwards edges already exist) but the idea of "just do Cython instead of NumPy" appears to make a fairly big performance difference.

Adding the hacky fix mentioned above now also fine; there's plenty of time to get a better fix in before 1.8, though, so might not be worth the overhead.

And not sure if the fix is worth backporting: The issue appears only whenever the graph has more than 2^16 edges. In most cases, the overflow will just cause a crash but it _is_ possible to avoid the crash and get wrong results instead.

**Comment by tupui** [comment]
Ok thanks, I will mark this for the next release.
